### PR TITLE
Respect culling in set_index if divisions are given

### PR DESCRIPTION
**Pull request by phofl**
ref #525

### CODE REVIEW

File: dask_expr/_shuffle.py [hunk 1]
```diff
@@ -1122,9 +1122,26 @@ def operation(df, index_name, drop, set_name, column_dtype):
...
+        if self.frame.npartitions < len(divisions) - 1:
+            part_filter = list(self.frame.find_operations(PartitionsFiltered))
+            if len(part_filter) > 0:
+                return tuple(
+                    [
+                        div
+                        for i, div in enumerate(divisions)
+                        if i in part_filter[0]._partitions
+                    ]
+                    + [divisions[-1]]
```
Comment (milesgranger) [inline dask_expr/_shuffle.py:1135]: So this `part_filter` is always length==1 or do we only ever care about the first element in the list?

File: dask_expr/tests/test_categorical.py [hunk 2]
```diff
@@ -48,3 +48,15 @@ def test_get_categories_simplify_adds_projection(df):
...
+
+
+def test_categorical_set_index():
+    df = lib.DataFrame({"x": [1, 2, 3, 4], "y": ["a", "b", "b", "c"]})
+    df["y"] = lib.Categorical(df["y"], categories=["a", "b", "c"], ordered=True)
+    a = from_pandas(df, npartitions=2)
+
+    b = a.set_index("y", divisions=["a", "b", "c"], npartitions=a.npartitions)
+    d1, d2 = b.get_partition(0), b.get_partition(1)
+    print(d1.index.optimize(fuse=False).pprint())
```
Comment (milesgranger) [inline dask_expr/tests/test_categorical.py:60]: left over debug print?

**Review by milesgranger — APPROVED** [review]
Looks generally okay to me, only a couple things/clarifications

File: dask_expr/_shuffle.py (hunk 1 above)
Comment (phofl) [reply dask_expr/_shuffle.py:1135]: We care about the first child that is filtered, you could imagine somtethong like this:


expr -> restrict to partitions 1-5 -> another expr -> another expr -> restrict to partitions 2 -> expr -> us

We only care about the filter to 2, the fact that 1-5 was filtered earlier is not relevant for us

File: dask_expr/tests/test_categorical.py (hunk 2 above)
Comment (phofl) [reply dask_expr/tests/test_categorical.py:60]: Yep, thx for catching

File: dask_expr/_shuffle.py (hunk 1 above)
Comment (milesgranger) [reply dask_expr/_shuffle.py:1135]: ah, I see. Thanks!

**Pull request merged**